connexion/decorators/validation.py

```diff
@@ -114,81 +211,28 @@ def __call__(self, function):
-
-                    if ctype_is_json:
-                        # Content-Type is json but actual body was not parsed
-                        return problem(400,
-                                       "Bad Request",
-                                       "Request body is not valid JSON"
-                                       )
-                    else:
-                        # the body has contents that were not parsed as JSON
-                        return problem(415,
-                                       "Unsupported Media Type",
-                                       "Invalid Content-type ({content_type}), expected JSON data".format(
-                                           content_type=request.headers.get("Content-Type", "")
-                                       ))
-
-                logger.debug("%s validating schema...", request.url)
-                error = self.validate_schema(data, request.url)
-                if error and not self.has_default:
-                    return error
-            elif self.consumes[0] in FORM_CONTENT_TYPES:
-                data = dict(request.form.items()) or (request.body if len(request.body) > 0 else {})
-                data.update(dict.fromkeys(request.files, ''))  # validator expects string..
-                logger.debug('%s validating schema...', request.url)
-
-                if self.strict_validation:
-                    formdata_errors = self.validate_formdata_parameter_list(request)
-                    if formdata_errors:
-                        raise ExtraParameterProblem(formdata_errors, [])
-
-                if data:
-                    props = self.schema.get("properties", {})
-                    errs = []
-                    for k, param_defn in props.items():
-                        if k in data:
-                            try:
-                                data[k] = coerce_type(param_defn, data[k], 'requestBody', k)
-                            except TypeValidationError as e:
-                                errs += [str(e)]
-                                print(errs)
-                    if errs:
-                        return problem(400, 'Bad Request', errs)
-
-                error = self.validate_schema(data, request.url)
+            content_handler = self.lookup_content_handler(request)
+            exact_match = request.content_type in self.consumes
+            partial_match = content_handler and content_handler.name in self.consumes
+            if not (exact_match or partial_match):
+                msg = "Invalid Content-type ({content_type}), expected one of {consumes}"
+                msg = msg.format(content_type=request.content_type, consumes=self.consumes)
+                return problem(415, "Unsupported Media Type", msg)
+
+            if content_handler:
+                error = content_handler.validate(request)
                 if error:
                     return error
 
+            elif self.strict_validation:
```

Comment:
If this is a global setting it means that if any endpoints accept non handled content types (e.g. binary uploads) then strict validation would have to be disabled for all endpoints.

Reply:
Yeah - I'm still thinking through the implications. The `strict` flag seems pretty nebulous. AFAICT it mostly checks for extra form fields, and errors if those exit.

So far this diff makes connexion much more strict, which I personally think is a good thing, but others might disagree (and it may cause breakage for people that have been relying on connexion's past flexibility).

I think personally as a user I'd prefer to explicitly have a content handler for every content type that I accept. Part of my goal is to make it straightforward to register new content handlers.

We could also provide more 'batteries included' beyond just json and formdata. For example, I could add a default octet stream content handler.

Reply:
Oh right I see. If it is straight forward to add a custom content handler that would be ideal for my case in #592 - especially if the custom handler/validator can process as a stream.